libtenzir/include/tenzir/import_stream.hpp

```diff
@@ -0,0 +1,218 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2023 The Tenzir Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#pragma once
+
+#include <tenzir/node_control.hpp>
+#include <tenzir/plugin.hpp>
+#include <tenzir/table_slice.hpp>
+
+#include <caf/attach_stream_source.hpp>
+#include <caf/scheduled_actor.hpp>
+#include <caf/scoped_actor.hpp>
+#include <caf/typed_event_based_actor.hpp>
+
+#include <deque>
+
+namespace tenzir {
+
+namespace detail {
+
+struct import_source_state {
+  bool stop;
+  std::deque<table_slice> queue;
+};
+
+class import_source_driver final
+  : public caf::stream_source_driver<
+      caf::broadcast_downstream_manager<table_slice>> {
+public:
+  explicit import_source_driver(std::shared_ptr<import_source_state> state,
+                                std::function<void(caf::error)> on_done)
+    : state_{std::move(state)}, on_done_{std::move(on_done)} {
+  }
+
+  void pull(caf::downstream<table_slice>& out, size_t num) override {
+    auto processed = size_t{0};
+    while (processed < std::min(num, state_->queue.size())) {
+      auto& slice = state_->queue[processed];
+      if (slice.rows() > 0) {
+        slice.import_time(time::clock::now());
+        out.push(std::move(slice));
+      }
+      processed += 1;
+    }
+    state_->queue.erase(state_->queue.begin(),
+                        state_->queue.begin()
+                          + detail::narrow<ptrdiff_t>(processed));
+  }
+
+  auto done() const noexcept -> bool override {
+    return state_->queue.empty() && state_->stop;
+  }
+
+  void finalize(const caf::error& error) override {
+    if (error && error != caf::exit_reason::unreachable) {
+      on_done_(error);
+    } else {
+      on_done_({});
+    }
+  }
+
+private:
+  std::shared_ptr<import_source_state> state_;
+  std::function<void(caf::error)> on_done_;
+};
+
+} // namespace detail
+
+// TODO: This is probably not the right abstraction.
+class import_stream {
```

Comment:
I'm now pretty sure that is a bad abstraction. Shall we proceed anyway and revisit this once there is more clarity regarding how this should work?